File: frontend/src/lib/journal.ts
import { authenticatedApiRequest } from './api'

export interface JournalEntry {
  id: string
  text: string
  source: 'journal'
  created_at: string
}

interface JournalListResponse {
  items: JournalEntry[]
  count: number
}

export function createJournal(text: string): Promise<JournalEntry> {
  return authenticatedApiRequest<JournalEntry>('/journal', {
    method: 'POST',
    body: JSON.stringify({ text }),
  })
}

export async function listJournals(limit = 20): Promise<JournalEntry[]> {
  const response = await authenticatedApiRequest<JournalListResponse>(
    `/journal?limit=${limit}`,
  )
  return response.items
}

File: backend/app/models/journal.py
"""Journal API request and response models."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


class JournalCreate(BaseModel):
    """Validated journal submission."""

    text: str = Field(min_length=1, max_length=10_000)

    @field_validator("text")
    @classmethod
    def reject_blank_text(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("Journal text cannot be blank")
        return normalized


class JournalRecord(BaseModel):
    """Journal entry returned to its owner."""

    id: str
    text: str
    source: Literal["journal"] = "journal"
    created_at: datetime


class JournalListResponse(BaseModel):
    """Recent journal entries for the authenticated user."""

    items: list[JournalRecord]
    count: int
